pkg/debug/common/common.go

```diff
@@ -0,0 +1,174 @@
+	for {
+		err := yaml.Unmarshal([]byte(yamlStr), &resultMap)
+		if err == nil {
+			break
+		}
+		errStr := err.Error()
+		r, _ := regexp.Compile("(?:line )(.*)(?::)")
+		lineIndex := r.FindStringSubmatch(errStr)[1]
+		lineIndexInt, err := strconv.Atoi(lineIndex)
+		if err != nil {
+			return "", fmt.Errorf("could not avoid error to parse yaml")
+		}
+		rm := strings.Split(yamlStr, "\n")[lineIndexInt-1]
+		yamlStr = string(bytes.Replace([]byte(yamlStr), []byte(rm+"\n"), []byte(""), 1))
+	}
+	for i := 0; i < len(str)-1; i++ {
+		var ok bool
+		resultMap, ok = resultMap[str[i]].(map[string]interface{})
+		if !ok {
+			return "", fmt.Errorf("wrong hierarchy: %s", str[i])
+		}
+	}
+	ret, ok := resultMap[str[len(str)-1]].(string)
+	if !ok {
+		return "", fmt.Errorf("wrong hierarchy: %s", str[len(str)-1])
+	}
+	return ret, nil
+}
+
+func ExtractFromGet(str string, column string) (string, error) {
+	column += " " // avoid only get prefix
+	lines := strings.Split(str, "\n")
+	nodeIndex := strings.Index(lines[0], column)
+	if nodeIndex == -1 {
+		return "", fmt.Errorf("could not found column: %s", column)
+	}
+	if len(lines) <= 1 || len(lines[1]) < nodeIndex {
+		return "", fmt.Errorf("could not found column: %s", column)
+	}
+	nodeName := strings.Split(string(lines[1][nodeIndex:]), " ")[0]
+	return nodeName, nil
+}
+
+func Debug(chaosType string, chaos string, ns string) ([]string, error) {
+	chaosType = strings.ToLower(chaosType)
+	cmd := fmt.Sprintf("kubectl get %s -n %s", chaosType, ns)
+	out, err := exec.Command("bash", "-c", cmd).CombinedOutput()
+	if err != nil {
+		return nil, fmt.Errorf("run command '%s' failed with: %s", cmd, err.Error())
+	}
+
+	lines := strings.Split(string(out), "\n")
+	var chaosList []string
+	if lines[1] == "" {
+		fmt.Printf(string(out))
+	} else {
+		title := lines[0]
+		chaosNum := 0
+		for i := 1; i < len(lines)-1; i++ {
+			chaosName, err := ExtractFromGet(title+"\n"+lines[i], "NAME")
+			if err != nil {
+				return nil, fmt.Errorf("extractFromGet failed with: %s", err.Error())
+			}
+			if chaos == "" || chaos == chaosName {
+				chaosList = append(chaosList, chaosName)
+				chaosNum++
+			}
+		}
+		if chaosNum == 0 {
+			return nil, fmt.Errorf("no chaos is found, please check your input")
+		}
+	}
+	return chaosList, nil
+}
+
+func GetPod(chaosType string, chaos string, ns string) (*PodName, error) {
+	// get podName
+	cmd := fmt.Sprintf("kubectl describe %s %s -n %s", chaosType, chaos, ns)
+	out, err := exec.Command("bash", "-c", cmd).CombinedOutput()
+	if err != nil {
+		return nil, fmt.Errorf("run command '%s' failed with: %s", cmd, err.Error())
+	}
+
+	failedMessage := regexp.MustCompile("(?:Failed Message:\\s*)(.*)").FindStringSubmatch(string(out))
+	if len(failedMessage) != 0 {
+		return nil, fmt.Errorf("chaos failed with: %s", failedMessage)
+	}
+	podHier := []string{"Status", "Experiment", "Pod Records", "Name"}
+	podName, err := ExtractFromYaml(string(out), podHier)
```

Comment:
Oops. Please use kubernetes client to get pod or any needed resources :crying_cat_face:, but not through `kubectl` commands.

Reply:
Hi @YangKeao I just rewrite the shell-wise part with controller-runtime/client-go. Welcome review~